src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/OperationInvokerBehavior.cs

```diff
@@ -38,27 +76,60 @@ void IOperationBehavior.ApplyDispatchBehavior(OperationDescription description,
 
             if (description.TaskMethod != null)
             {
-                dispatch.Invoker = new TaskMethodInvoker(_serviceProvider, description.TaskMethod, description.TaskTResult);
+                if (s_isOperationInvokerGeneratorEnabled.Value)
+                {
+                    if (DispatchOperationRuntimeHelpers.OperationInvokers.TryGetValue(DispatchOperationRuntimeHelpers.GetKey(description.TaskMethod), out IOperationInvoker invoker))
+                    {
+                        dispatch.Invoker = invoker;
+                    }
+                    else
+                    {
+                        // reaching this point means that the operation invoker generator is enabled but we did not succeed to fetch the invoker
+                        // for the operation. This is a bug in the DispatchOperationRuntimeHelpers.GetKey logic.
+                        throw new PlatformNotSupportedException();
+                    }
+                }
+                else
+                {
+                    dispatch.Invoker = new TaskMethodInvoker(_serviceProvider, description.TaskMethod, description.TaskTResult);
+                }
+
             }
             else if (description.SyncMethod != null)
             {
-                if (description.BeginMethod != null)
+                if (s_isOperationInvokerGeneratorEnabled.Value)
                 {
-                    // both sync and async methods are present on the contract, check the preference
-                    //OperationBehaviorAttribute operationBehaviorAttribute = description.Behaviors.Find<OperationBehaviorAttribute>();
-                    //if ((operationBehaviorAttribute != null) && operationBehaviorAttribute.PreferAsyncInvocation)
-                    //{
-                    //    dispatch.Invoker = new AsyncMethodInvoker(description.BeginMethod, description.EndMethod);
-                    //}
-                    //else
-                    //{
-                    dispatch.Invoker = new SyncMethodInvoker(_serviceProvider, description.SyncMethod);
-                    //}
+                    if (DispatchOperationRuntimeHelpers.OperationInvokers.TryGetValue(DispatchOperationRuntimeHelpers.GetKey(description.SyncMethod), out IOperationInvoker invoker))
+                    {
+                        dispatch.Invoker = invoker;
+                    }
+                    else
+                    {
+                        // reaching this point means that the operation invoker generator is enabled but we did not succeed to fetch the invoker
+                        // for the operation. This is a bug in the DispatchOperationRuntimeHelpers.GetKey logic.
+                        throw new PlatformNotSupportedException();
```

Comment:
It could also be that the contract lives in a separate assembly from the main application assembly and generation of the invoker wasn't enabled in that other assembly. I think we should throw an InvalidOperationException and have a string resource saying if the contract interface is in a different assembly than the application, generation of the operation invoker must be enabled when compiling the contract interface assembly. This is a lot to describe in a string resource. I wonder if we should have a KB section at [https://corewcf.github.io/kb/[num](https://corewcf.github.io/kb/[num)] where we could document how to handle a particular error case. We shouldn't block on that for this PR, but it's something I think we should do.

Reply:
created #1417 to track creation of the KB section at corewcf.github.io